src/test/groovy/com/example/demo/LocationTests.groovy

```diff
@@ -0,0 +1,192 @@
+        pageableResult.content[0].getLocation() == loc3
+        pageableResult.content[1].getLocation() == loc2
+        pageableResult.content[2].getLocation() == loc1
+
+    }
+
+    def "LocationController - REST - adding valid location test"() {
+        given: "an entity"
+        when: "a valid call to addLocation is made"
+        def result = mockMvc.perform(put("/api/location/addLocation?location=Cork")).andReturn().response.contentAsString
+        def json = new JsonSlurper().parseText(result)
+        then: "result is an entity with an id"
+        json.id == 1
+        json.id > 0
+    }
+
+    def "LocationController - REST - adding invalid location test"() {
+        given: "an entity"
+        def e1 = locationRepository.save(new LocationEntity("Cork"))
+        when: "an invalid call to addLocation is made"
+        def except = mockMvc.perform(put("/api/location/addLocation?location=Cork")).andReturn().getResolvedException().getClass()
+        then: "response status exception thrown"
+        assert(except.is(ResponseStatusException))
+    }
+
+    def "LocationController - REST - valid findById"() {
+        given: "an entity"
+        def loc = "Cork"
+        def e1 = locationRepository.save(new LocationEntity(loc))
+        when: "a valid call to addLocation is made"
+        def result = mockMvc.perform(get("/api/location/findById?id=1")).andReturn().response.contentAsString
+        def json = new JsonSlurper().parseText(result)
+        then: "result is an entity with an id"
+        json.id == 1
+        json.location == loc
+    }
+
+    def "LocationController - REST - invalid findById"() {
+        given: "an entity"
+        def loc = "Cork"
+        def e1 = locationRepository.save(new LocationEntity(loc))
+        when: "a valid call to addLocation is made"
+        def except = mockMvc.perform(get("/api/location/findById?id=2")).andReturn().getResolvedException().getClass()
+        then: "response status exception thrown"
+        assert(except.is(ResponseStatusException))
+    }
+
+    def "LocationController - REST - valid deleteById"() {
+        given: "an entity"
+        def loc = "Cork"
+        def e1 = locationRepository.save(new LocationEntity(loc))
+        when: "a valid call to deleteById is made"
+        def result = mockMvc.perform(delete("/api/location/deleteById?id=1")).andReturn().response.contentAsString
+        then: "result verifies the delete"
+        result == "location deleted successfully."
+    }
+
+    def "LocationController - REST - invalid deleteById"() {
+        given: "an entity"
+        def loc = "Cork"
+        def e1 = locationRepository.save(new LocationEntity(loc))
+        when: "an invalid call to deleteById is made"
+        def except = mockMvc.perform(delete("/api/location/deleteById?id=2")).andReturn().getResolvedException().getClass()
+        then: "response status exception thrown"
+        assert(except.is(ResponseStatusException))
```

Comment:
We are probably more interested in the HTTP code here than the exception underlying it. Normal REST clients can't see that exception, so if we wanted to check further we could look at the json body